tests/pipelines/test_api.py

```diff
@@ -0,0 +1,207 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint:disable=redefined-outer-name
+# pylint:disable=unused-argument
+
+import os
+import copy
+import mock
+import pytest
+
+import databricks_cli.pipelines.api as api
+from databricks_cli.pipelines.api import LibraryObject
+
+PIPELINE_ID = '123456'
+SPEC = {
+    'id': PIPELINE_ID,
+    'name': 'test_pipeline'
+}
+CREDENTIALS = 'dummy_credentials'
+HEADERS = 'dummy_headers'
+
+
+@pytest.fixture()
+def pipelines_api():
+    with mock.patch('databricks_cli.pipelines.api.DeltaPipelinesService') \
+            as DeltaPipelinesServiceMock:
+        DeltaPipelinesServiceMock.return_value = mock.MagicMock()
+        _pipelines_api = api.PipelinesApi(None)
+        yield _pipelines_api
+
+
+def file_exists_stub(_, dbfs_path):
+    exist_mapping = {
+        'dbfs:/pipelines/code/40bd001563085fc35165329ea1ff5c5ecbdbbeef.jar': True,  # sha1 of 123
+        'dbfs:/pipelines/code/51eac6b471a284d3341d8c0c63d0f1a286262a18.jar': False,  # sha1 of 456
+    }
+    return exist_mapping[dbfs_path.absolute_path]
+
+
+@mock.patch('databricks_cli.dbfs.api.DbfsApi.file_exists', file_exists_stub)
+@mock.patch('databricks_cli.dbfs.dbfs_path.DbfsPath.validate')
+@mock.patch('databricks_cli.pipelines.api.PipelinesApi._get_credentials_for_request')
+@mock.patch('databricks_cli.dbfs.api.DbfsApi.put_file')
+def test_deploy(put_file_mock, get_credentials_mock, dbfs_path_validate, pipelines_api, tmpdir):
+    """
+    Scenarios Tested:
+    1. All three types of local file paths (absolute, relative, file: scheme)
+    2. File already present in dbfs
+    3. Local files already present in dbfs
+    4. Local files that need to be uploaded to dbfs
+    Every test local file which has '123' written to it is expected to be already present in Dbfs.
+    A test local file which has '456' written to it is not present in Dbfs and therefore must be.
+    uploaded to dbfs.
+    """
+    get_credentials_mock.return_value = CREDENTIALS
+    deploy_mock = pipelines_api.client.client.perform_query
+    # set-up the test
+    jar1 = tmpdir.join('/jar1.jar').strpath
+    jar2 = tmpdir.join('/jar2.jar').strpath
+    jar3 = tmpdir.join('/jar3.jar').strpath
+    jar4 = tmpdir.join('/jar4.jar').strpath
+    jar3_relpath = os.path.relpath(jar3, os.getcwd())
+    jar4_file_prefix = 'file:{}'.format(jar4)
+    with open(jar1, 'w') as f:
+        f.write('123')
+    with open(jar2, 'w') as f:
+        f.write('456')
+    with open(jar3, 'w') as f:
+        f.write('456')
+    with open(jar4, 'w') as f:
+        f.write('456')
+    libraries = [{'jar': 'dbfs:/pipelines/code/file.jar'},
+                 {'jar': jar1},
+                 {'jar': jar2},
+                 {'jar': jar3_relpath},
+                 {'jar': jar4_file_prefix}]
+    SPEC['libraries'] = libraries
```

Comment:
this looks really weird. Why are we mutating an module level object which looks like a constant?

Reply:
I agree it looks weird, I'm switching to deepcopying the SPEC to spec and then adding the libraries field to it.